Fixed #15053 -- Improved template extending algorithm.

prestontimmons opened the pull request.

timgraham commented:
Now that Aymeric's branch has merged, are we still targeting this for 1.8? I can help with review if so.

prestontimmons commented:
Hi Tim,

I think this will be post-1.8.

There are some details still up in the air about the origin api. My changes
will formalize this to an extent, but it will need further definition for
multiple template engines.

With that said, I still hope to push an update soon that accounts for the
discussion I had with Aymeric on the mailing list. I was delayed a bit by
getting sick last week.

On Wed, Jan 7, 2015 at 5:58 AM, Tim Graham <notifications@github.com> wrote:

> Now that Aymeric's branch has merged, are we still targeting this for 1.8?
> I can help with review if so.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/django/django/pull/3475#issuecomment-69012741>.
>

timgraham commented on docs/ref/templates/api.txt:
```diff
@@ -910,38 +910,189 @@ option. It uses each loader until a loader finds a match.
+        the template source. A filesystem-based loader may return the full
+        path to the file as the ``template_origin``, for example.
+
+        .. versionchanged:: 1.9
```
.. deprecated:: 1.9
(in which case you can probably drop the first sentence)

timgraham commented on docs/releases/1.9.txt:
```diff
@@ -243,6 +248,16 @@ a directory. Now, Django only silences the exception if the template source
 does not exist. All other situations result in the original ``IOError`` being
 raised.
 
+Template LoaderOrigin and StringOrigin are removed
```
`` around class names

timgraham commented:
Will try to give this a full review on Monday.

prestontimmons commented:
Thanks, Tim. I updated and rebased to pull in the latest changes.

prestontimmons commented on django/template/loader.py:
```diff
@@ -4,25 +4,20 @@
 
```
The tests for the changes in this file are in ``tests/template_loader/tests.py``.

prestontimmons commented on django/template/loader_tags.py:
```diff
@@ -92,6 +93,43 @@ def __init__(self, nodelist, parent_name, template_dirs=None):
+                    raise ExtendsError(
+                        "Cannot extend templates recursively when using "
+                        "non-recursive template loaders",
+                    )
```
This check is what fixes #22232 when non-recursive loaders are being used.

prestontimmons commented on tests/template_tests/test_loaders.py:
```diff
@@ -10,8 +10,9 @@
 
```
For each loader I added tests for ``Loader.get_template`` and ``Loader.load_template_source``. I felt the latter was necessary since otherwise Django would no longer exercise those code paths through the deprecation period.

prestontimmons commented on django/template/base.py:
```diff
@@ -154,23 +162,30 @@ class InvalidTemplateLibrary(Exception):
+        if self.loader:
+            return '%s.%s' % (
+                self.loader.__module__, self.loader.__class__.__name__,
+            )
```
If a non-Django engine was to provide postmortem information, we'd need to provide a base ``Origin`` class. The only required attributes are ``name`` and ``loader_name``, where ``name`` is the template identifier and ``loader_name`` is a string identifying the method used to loader the template.

Jinja2 doesn't look like it exposes that information, though. So I just kept things in ``django.template.base`` for now.

aaugustin replied:
If I understand correctly, the class is super easy to duck-type. This reduces the need for a base class.

prestontimmons commented on django/template/base.py:
```diff
@@ -131,7 +131,15 @@ class TemplateSyntaxError(Exception):
+    """
+    def __init__(self, msg, tried=None):
+        self.tried = tried or []
+        super(TemplateDoesNotExist, self).__init__(msg)
```
Bikeshedding: I think ``TemplateDoesNotExist`` makes more sense in ``django.template.loader`` than ``django.template.base``. We can't move it yet, though, until ``django.template.loader.BaseLoader`` is removed. Otherwise it leads to circular imports.

aaugustin replied:
Even better, `django.template.exceptions`?

Not a big deal anyway, do what you want.

prestontimmons commented on django/views/debug.py:
```diff
@@ -266,61 +265,33 @@ def __init__(self, request, exc_type, exc_value, tb, is_email=False):
+                        entry[0].loader.engine == e
+                    ],
+                ))
+            self.postmortem = postmortem
```
There's one case this information isn't quite correct. If an ``Engine`` is manually created and not part of ``settings.TEMPLATES``, the debug page won't know about it. Is that case worth accounting for?

aaugustin replied:
This doesn't matter much. Just ensure that such edge cases don't raise exceptions.

prestontimmons commented:
One more thing, there's a sample project to accompany these changes:

https://github.com/prestontimmons/project-15053

It has a multiple engine configuration setup so you can see how the postmortem appears for various situations in the debug page.

timgraham commented on docs/ref/templates/api.txt:
```diff
@@ -922,38 +922,189 @@ option. It uses each loader until a loader finds a match.
+
+        .. warning::
+
+            The ``template_dirs`` argument has been deprecated. Loaders that
```
I couldn't find any details of this deprecation.

prestontimmons replied:
I might be wrong here. I thought ``template_dirs`` was implicitly deprecated based on:

https://docs.djangoproject.com/en/dev/releases/1.8/#dirs-argument-of-template-finding-functions

We can keep it if necessary. It still has the problem of not carrying through to the include and extends tag, though.

timgraham replied:
If it's only needed for the deprecation cycle, then I suggest we just not document it in these new methods.

prestontimmons replied:
I'm okay with not documenting it, but I'll need to special-case the filesystem and cached loaders to conditionally pass ``template_dirs`` to ``get_template_sources``. Something like this:

https://dpaste.de/itr6

This way, custom loaders that don't specify ``template_dirs`` won't lead to an error.

aaugustin replied:
Since you're introducing a new API and the deprecated argument is just there for backwards-compatibility within Django, I agree that you shouldn't document it.

Django already uses `getargspec` in a few places to account for the removal of arguments from documented APIs.

timgraham commented on django/template/loaders/base.py:
```diff
@@ -44,3 +93,10 @@ def reset(self):
+    @property
+    def supports_recursion(self):
+        """
+        Used during deprecation period of non-recursive loaders.
```
Don't see details about this deprecation.

prestontimmons replied:
Non-recursive loaders are ones that only define ``load_template_sources``. The deprecation is noted here:

https://github.com/django/django/pull/3475/files#diff-c0f4f3a0c26f12ebc2e41904af710f4dR361

Do you think this just needs a clearer doc string? Or is there something else I should add?

prestontimmons replied:
Nevermind, I see what you mean now about the ``supports_recursion`` deprecation. I'll update it.

prestontimmons replied:
Since ``supports_recursion`` is called whenever extending happens, adding a ``DeprecatedInDjango21Warning`` to ``supports_recursion`` breaks a large number of Django tests, not all related to templates. Is it sufficient to add a note to ``internals/deprecations.txt``? Or else add the warning filter to ``runtests.py``?

timgraham replied:
I'm fine with whatever you think is best. I think the number of people writing custom template loaders is low, so docs might be sufficient if a warning will be more effort.

timgraham commented on django/views/debug.py:
```diff
@@ -907,12 +884,14 @@ def default_urlconf(request):
-{% for t in loader.templates %}{{ t.name }} ({{ t.status }})
-{% endfor %}{% endfor %}
-{% else %}Django couldn't find any templates because your 'loaders' option is empty!
+{% if template_does_not_exist %}Template loader postmortem
```
If you wanted to use indentation to try to add some clarity to this, I wouldn't be opposed.

prestontimmons replied:
Sure. Do you mean indentation in the template code or indentation in the output? Whitespace gets output verbatim since this goes into a textarea, Some of the existing output displays poorly because of this. For instance, here's output from a ``TemplateSyntaxError``:

```
  1 : {% load badtags %}



   2 : 



   3 : {% block content %}



   4 :   Hello



   5 :   {{ a }}



   6 :    {{ coconuts }} 



   7 : {% endblock %}



   8 : 
```

Fixing that is possible but requires making the template code harder to read.

A couple thoughts:

1) If filters could be registered via the ``Engine``, I think a custom filter could be used here to collapse whitespace and get both a readable template and output.

2) Is there any reason not to reuse the ``TECHNICAL_500_TEXT_TEMPLATE`` for the copy/paste text box? That would make things a bit simpler, also.

timgraham replied:
Oh, didn't realize it was so complicated. We can save it for a follow up commit.

timgraham commented on tests/template_tests/test_loaders.py:
```diff
@@ -32,26 +34,55 @@ def create_engine(self, **kwargs):
-        """
-        engine = self.create_engine()
+    @contextmanager
+    def set_debug(self):
```
used?

prestontimmons replied:
The diff is unclear here. This test is moved down intact to line 98.

timgraham replied:
comment was about the `set_debug` function -- don't see another mention of it in this diff.

timgraham commented:
suggested edits: http://dpaste.com/2Z46ATK
Will wait for Aymeric's review before merge.

xrmx commented:
Il 27/03/2015 13:58, Tim Graham ha scritto:
> suggested edits: http://dpaste.com/2Z46ATK
> Will wait for Aymeric's review before merge.

The paste looks expired

timgraham commented:
The original one had a bug, so I made a new one and edited my comment.

aaugustin commented on django/template/base.py:
```diff
@@ -131,7 +131,15 @@ class TemplateSyntaxError(Exception):
+    This exception is used when template loaders are unable to find a
+    template. The tried argument is an optional list of tuples containing
+    (origin, status), where origin is an Origin object and status is a string
+    with the reason the template wasn't found.
```
I like this API :+1:

aaugustin commented on django/template/loaders/cached.py:
```diff
@@ -16,15 +16,86 @@ class Loader(BaseLoader):
+            )
+        except TemplateDoesNotExist as e:
+            self.get_template_cache[key] = e
+            raise e
```
Unless you did this on purpose, a simple `raise` is better, as it avoids a needless frame in the traceback.

aaugustin commented on django/template/loaders/cached.py:
```diff
@@ -16,15 +16,86 @@ class Loader(BaseLoader):
-            return '-'.join([template_name, hashlib.sha1(force_bytes('|'.join(template_dirs))).hexdigest()])
-        else:
-            return template_name
+            return '-'.join([value, hashlib.sha1(force_bytes('|'.join(template_dirs))).hexdigest()])
```
`value` could get quite long here. What about a key in the format `<template name>-<hash of skip>-<hash of template_dirs>`?

aaugustin commented on django/views/debug.py:
```diff
@@ -266,61 +265,33 @@ def __init__(self, request, exc_type, exc_value, tb, is_email=False):
+                if hasattr(engine, 'engine'):
+                    e = engine.engine
+                else:
+                    e = engine
```
I'm concerned by this bit. Basically you're pulling the underlying `django.template.engine.Engine` from the `django.template.backends.django.DjangoTemplates` engine.

To support any template engine, it looks like you need to add an API on the template backend engine — that is, `DjangoTemplates` or its equivalent for another language.

If you're puposefully taking a shortcut, can you leave a comment explaining it? You're not making the situation worse anyway, considering that I left exactly that kind of shortcuts in code you're removing.

prestontimmons replied:
Yes, this part in the debug view is a shortcut. It's also the reason I haven't documented non-Django engine support yet. I don't mind fixing it, but it requires fleshing out the API further. Below is a possible solution. Does this make ``TemplateDoesNotExist`` too complicated?

My goal is to display a message like this if ``loader.get_template`` fails:

```
Django tried loading these templates, in this order:

Using engine django:
* django.template.loaders.filesystem.Loader: /path/to/template.html (Source does not exist)
* django.template.loaders.filesystem.Loader: /other/path/to/template.html (Source does not exist)

Using engine django2:
* django.template.loaders.filesystem.Loader: /another/path/to/template.html (Source does not exist)

Using engine jinja2:
This engine did not provide a list of tried templates.
```

The idea is to have each backend annotate ``TemplateDoesNotExist`` with the backend instance. Then the ``django.template.loader`` functions can raise a final ``TemplateDoesNotExist`` that contains all the intermediate exceptions.

For example, ``DjangoTemplates.get_template`` would look like this:

```python
def get_template(self, template_name, dirs=_dirs_undefined):
    try:
        return Template(self.engine.get_template(template_name, dirs))
    except TemplateDoesNotExist as e:
        e.backend = self
        raise
```

And ``django.template.loader`` would look like this:

```python
chain = []
for engine in engines:
    try:
        ...
    except TemplateDoesNotExist as e:
        chain.append(e)
raise TemplateDoesNotExist(chain=chain)
```

The debug code would be reduced to a simple loop through ``e.chain``. This would give access to each intermediate ``e.backend`` and ``e.tried``.

Note, if the exception happened during the render phase, the postmortem would only show information from the current active engine.

aaugustin replied:
It would seem more elegant to me to re-raise a new `TemplateDoesNotExist(..., backend=self.name)` than just annotate the current `TemplateDoesNotExist` with `e.backend = self`.

In `get_template()`, non-Django backends catch their own "template does not exist" exception and re-raise Django's standard `TemplateDoesNotExist`. The Django backend is a special case because it can re-use the same exception, but really that's an accident. I had to keep the same exception for the "Django templates library" and "Django templates backend" in order to preserve backward-compatibility.

The general idea makes sense to me. For now you could simply keep that comment:

    # TODO: add support for multiple template engines (#24120).
    # TemplateDoesNotExist should carry all the information.

and leave these changes for another pull request.

aaugustin commented on docs/ref/templates/api.txt:
```diff
@@ -922,38 +922,189 @@ option. It uses each loader until a loader finds a match.
+        This is where a filesystem loader would read contents from the
+        filesystem, or a database loader would read from the database. If a
+        matching template doesn't exist, this should raise a
+        ``TemplateDoesNotExist`` error.
```
Isn't the `LoaderOrigin` supposed to have validated that the template exists already?

prestontimmons replied:
Oops. I need to update these docs since ``LoaderOrigin`` is just ``Origin`` now.

``get_template_sources`` returns a list of possible sources rather than only valid, readable sources. This is for two reasons:

1) The postmortem information is collected in ``get_template``. If ``get_template_sources`` doesn't yield all sources, including invalid ones, we wouldn't be able to show debug information for the ones that don't exist, e.g. "Django tried loading these templates in this order..."

2) There's an edge case if a file or source is deleted between the call to ``get_template_sources`` and ``get_contents``. This would result in a system exception rather than ``TemplateDoesNotExist``.

aaugustin commented on docs/ref/templates/api.txt:
```diff
@@ -922,38 +922,189 @@ option. It uses each loader until a loader finds a match.
+        Returns a ``Template`` object for a given ``template_name`` by looping
+        through results from ``get_template_sources`` and calling
+        ``get_contents``. This returns the first matching template. If no
+        template is found, ``TemplateSyntaxError`` is raised.
```
That should be `TemplateDoesNotExist`, shouldn't it?

aaugustin commented on docs/ref/templates/api.txt:
```diff
@@ -922,38 +922,189 @@ option. It uses each loader until a loader finds a match.
+        .. warning::
+
+            The ``template_dirs`` argument has been deprecated. Loaders that
+            depend on this argument should be updated to no longer use it.
```
As above.

aaugustin commented on docs/ref/templates/api.txt:
```diff
@@ -922,38 +922,189 @@ option. It uses each loader until a loader finds a match.
+-------
+
+A simple loader is one that loads templates from a Python dictionary. An
+implementation of such a loader would look like:
```
That's basically the `LocMem` loader. It is worth repeating the code in the docs? What about suggesting to look at the implementation of built-in loaders instead?

aaugustin commented:
This is almost ready to land :-)

The only comment that really matters in my review is the one that starts with "I'm concerned by this bit.". 

Caveats:

- It isn't clear to me how a third-party engine can populate the new debug page. (I saw you couldn't extract interesting information from Jinja2.)
- I didn't look at tests.

prestontimmons commented:
Thanks for all the feedback. There are two ways another template engine could provide a postmortem, both of which happen at the backend level:

1) If the underlying engine loaders provide the list of tried templates in their internal exception, the backend can pass that information when ``TemplateDoesNotExist`` is reraised.

2) If the underlying engine exposes an equivalent to ``get_template_sources``, the backend can lookup the sources and pass them to ``TemplateDoesNotExist``.

I couldn't find a faculty for either 1 or 2 in Jinja2.

Since we have the dummy backend, I'll update that so we at least have one other engine using the postmortem API.

aaugustin commented:
Oh right, everything works with `TemplateDoesNotExist` (and that's why I :+1: the API).

Did you document it besides the docstring, to explain how other backends can take advantage of it? If you did, I missed it.

prestontimmons commented:
I think I've incorporated all the feedback in the branch. A few things:

* I added additional deprecation warnings and notes.

* I removed ``template_dirs`` from the docs and updated the base and cached loader to use ``getargspec`` for compatibility.

* I fixed ``LoaderOrigin`` references in the docs and replaced the example loader with a note to look at the built-in ones for examples.

I left these things for followup commits:

* Make template backends reraise ``TemplateDoesNotExist`` with a ``backend`` argument. This will enable a consistent api across all engines for the debug postmortem.

* Clean up the copy/paste box formatting and output.

timgraham commented:
Looks about ready to me. Aymeric, do you want to give it another pass?

prestontimmons commented:
Thanks, Tim. I made these updates. Good idea on adding ``RemovedInDjango21Warning`` comments. I added that to each place it's not explicitly raised.

timgraham commented on django/template/loaders/eggs.py:
```diff
@@ -20,19 +31,42 @@ def __init__(self, engine):
+    def get_contents(self, origin):
+        try:
+            source = resource_string(origin.app_name, origin.pkg_name)
+        except:
```
bare except is okay here?

prestontimmons replied:
Yeah, the bare except isn't optimal. That's behavior I carried over from the existing egg loader.

I didn't change it since I don't know what exceptions are raised with real eggs. The test suite isn't helpful here since we use a mock egg provider. Do you want me to look into that? Or can we make it a separate ticket for now?

timgraham replied:
separate ticket

timgraham commented on tests/template_tests/test_loaders.py:
```diff
@@ -256,10 +324,44 @@ def setUp(self):
+
+class LocmemLoaderTests(SimpleTestCase):
+
+    def setUp(self):
```
use `setUpClass()` as discussed in other patch?

timgraham commented:
Not sure if you plan to squash commits, but I'd probably add "Refs #15053" in the other commits if not.

prestontimmons commented:
Cool. I added in your suggestions. I kept ``setUp`` rather than ``setUpClass`` in the ``CachedLoaderTests`` though to ensure the cache is cleared between each test method.

prestontimmons commented:
Oops. I forgot to call super in ``setUpClass`. I'll update that.

aaugustin commented:
@timgraham I'm fine with this patch provided @prestontimmons has read and perhaps taken into account my comments, I don't think another pass is needed.

prestontimmons commented:
@aaugustin Cool. I think this PR includes what's reasonable at the moment. There are some things I'm eager to finish--like proper debug support for non-Django engine--but I don't want to make the current PR any bigger.

timgraham commented:
Patch needs a rebase, but I marked the ticket as ready for checkin so please merge at your convenience.

prestontimmons commented:
Great. Thanks. It's merged now.

xrmx commented:
Thanks to everyone involved!
